### ISSUE TITLE
Paypal JS SDK not loading for some users

### ISSUE THREAD

**Issue by freddyhaddad** [paypal/paypal-js#461]
### Library used

direct script integration, as follows: https://www.paypal.com/sdk/js?client-id=[OUR-CLIENT-ID]&components=messages,buttons

### 🐞 Describe the Bug

We set up sentry to report when an external <script> tag failed to load (triggers an onerror event).
For 1% of our users, the SDK script tag is unable to load. It works for 99% of the users since we still get payments, but cannot reproduce locally.
Watching some affected user sessions show the user being frustrated as he cannot find the Paypal Smart Buttons on the page.
Moreover, window.paypal is not available right under the asynchronous script tag for these users.

All browsers are affected, desktop and mobile, Chrome/Safari/Edge

We use the Paypal JS SDK on 3 separate websites, and the issue happens on the 3 sites.

### 🔬 Minimal Reproduction

Cannot reproduce

### 😕 Actual Behavior

The direct Paypal SDK JS script tag triggers an "onerror" event, and window.paypal isn't available which triggers subsequent errors upon usage.

### 🤔 Expected Behavior

Expected Paypal JS SDK to be downloadable by all users' browsers.

**Comment by aduca98** [comment]
Also facing this issue.

**Comment by KthProg** [comment]
We've identified the root cause of this issue as likely being Safari security settings. Specifically it seems like disabling "Hide IP from trackers" on both mobile and desktop Safari causes the button to load successfully in sessions where it previously did not load. Some devs have said that it appears to be a CORS issue. I think it is a security issue where Safari sees the paypal sdk as a tracker script (maybe because it is loading so frequently), and thus hides the user's IP address, which leads to a failure in the script loading, I assume do to some backend requirement at PayPal to have the IP when loading the script.

**Comment by majamaki** [comment]
We've been experiencing this issue intermittently.  We use Gravity Forms with their PayPal add-on.  It is frustrating as this issue prevents the form from loading and the page from completely loading, thus breaking our forms and losing business.   Any insights on how to fix this or is just abandoning PayPal the best option?

**Comment by AndriusTelesoftas** [comment]
It looks like https://github.com/paypal/paypal-js/issues/561 Could also be related to this.
That Issue has a little more details on it's reproduction, maybe it could further help with it's investigation.

**Comment by cesarhuret** [comment]
bump

**Comment by nguyenkhoa0721** [comment]
bump

**Comment by scottie-33** [comment]
bump

**Comment by ctapang** [comment]
bump

**Comment by freddyhaddad** [comment]
bump

**Comment by ctapang** [comment]
I have studied the issue and I believe this is a CORS issue. The problem is partly in paypal-js code and parlty in the server code that distributes the SDK (https://www.paypal.com/sdk/js). General information about CORS related to this can be found here:
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS, but here are the details:

1. In the paypal-js package there is code that fetches the SDK. The header for this fetch must have an Origin record. The Origin must contain the url of the store server (the store selling things that can be paid using Paypal).
2. In the server response to the fetch, there must also be a header which should contain the record "Access-Control-Allow-Origin: *". Right now the server that serves the SDK does not include any header in its response.

**Comment by ctapang** [comment]
I fixed it by copying the Paypal client code from node_modules, and then modifying it such that it loads the SDK from our server instead of from the Paypal server. It was easy enough to download the SDK using curl (it can be downloaded even without specifying your client_id). I then placed that SDK in our server.

**Comment by AndriusTelesoftas** [comment]
@ctapang It does sound good that a workaround exists. But [Paypal docs](https://developer.paypal.com/sdk/js/configuration/) explicitly say: "Don't include it in a bundle or host it yourself". This issue already persists for about a year, I wish someone would tackle them already.

**Comment by ctapang** [comment]
Yes, I got the SDK code to download from our server, but when I finally made it to just display the PayPal buttons, none of the buttons work. It seems that the PayPal server is refusing to communicate with the same SDK but downloaded from our server.

**Comment by ctapang** [comment]
Paypal needs to fix this. The devices and browsers that "fail" (about 1% according to some who have reported this issue) are actually not failing. The big problem for Paypal is that this is a security hole in 99% of the cases that seem to work. They have to fix it soon.

**Comment by aceoft** [comment]
I also appear to be having this issue. Unfortunately, the PayPal SDK was loaded in the head of my document and was preventing entire pages from loading. In the network tab, the paypal script just sits with a spinner next to it.

**Comment by staabm** [comment]
we also see this issue

**Comment by SerafinDinges** [comment]
We also see this issue and can confirm that disabling the "Hide IP address" setting in safari settings fixes it. This means a huge load of our users can't access PayPal.

**Comment by SerafinDinges** [comment]
We are able to reproduce this on some Safari browsers on macOS too. Even more infuriating is that we have two browsers with the exact same Safari and macOS build numbers – on one of them the setting works fine on one the setting disables the PayPal button completely and we get an error when loading the script.

<img width="670" alt="image" src="https://github.com/user-attachments/assets/163ecba4-0abe-4beb-92ed-928f5264c76b">

**Comment by Khauru** [comment]
Seems like it's working again. We haven't changed anything

**Comment by BaddyDolly** [comment]
same issue on webflow hosted website that has PayPal enabled as part of its e-commerce offering. In this case the script is not loading for about 60-100secs. Same issue occurring with the stripe-integration btw.
![paypal2](https://github.com/user-attachments/assets/60a345a1-411b-4715-bfe8-bfb17365c157)
![paypalnotloading](https://github.com/user-attachments/assets/cf6bc427-ca6d-42e4-a618-6ef393c5564d)

**Comment by Some0ne7** [comment]
Same issue here for a webflow hosted website with PayPal enabled. The below are screenshots replicating the issue in the iOS 18 Xcode simulator (also real users complained). Interestingly same issue happening with the stripe-js occasionally not loading for 60-100 seconds.

![paypal2](https://github.com/user-attachments/assets/6c93688f-bbfd-46ea-aab0-e4f02014b545)
![paypalnotloading](https://github.com/user-attachments/assets/fb669b7f-cbd5-49a6-9111-0be836948ea2)

**Comment by Jaikant** [comment]
Not sure if this could be helpful to others:

I was seeing the same issue of:
<img width="1050" alt="Screenshot 2024-12-24 at 10 46 15 AM" src="https://github.com/user-attachments/assets/60e13b43-b024-407c-9094-8239afe2d798" />

I took the url in the console log and putting it on a browser and I could see the actual error message:
![Screenshot 2024-12-24 at 10 47 29 AM](https://github.com/user-attachments/assets/4a80df0b-5f25-4d47-941a-62c919af07a3)


It seems for some reason the client id has become invalid.

**Comment by ctapang** [comment]
bump

A couple of issues described above may be unrelated. It still looks like a CORS issue to me:
https://github.com/paypal/paypal-js/issues/461#issuecomment-2481152272

**Comment by ctapang** [comment]
bump

**Comment by tj-dixon** [comment]
This issue is not stale.
